challenges/advanced-callable/question.py

```diff
@@ -0,0 +1,46 @@
+"""
+TODO:
+
+Define a callable type that takes two arguments:
+
+1. *arbitrary param name*: a string
+2. `size`: an integer
+
+Returns: None
+
+HINT: Use Protocol
```

Comment:
Is this still true that we want to use `Protocol` here?

Reply:
Yeah, I split this challenges into 3 difficulty levels. Only the basic doesn't need Protocol

Reply:
Emm, it can pass without using Protocol

https://pyright-play.net/?code=FAMwTg9gtgBAxhANogpnALgSwgOwM4B0AhgEZwyZQAOEY6MAwkcqasMOgJ5UowCynJixKoYAXkbNErFAG1ZedGAA0FHOgC6qgHK4UG9gBMUIGHgAWEAK6JDAfSpE8eABQBKAFzAYPmMdPoKIp2IFY4cC6h4R78glIybjAAtAB8MLo4KF6%2BOTAE%2Bew5-jA4RFAodniYAF4oLqXlMYoqZjVZauiJqel62bk%2B%2BQSFvsVcPJVtLmMVTUqqVbUxmOpdaRlZ3v15BZs%2BgcFREQ0VCyhuuzD76CFhEdMTtedGJmaWNvYgRJiI7n0%2Bo0FrodIrcYgIhNIRGdkmtehccoNhv8XkQAOYnSZo9rLdDzNpLFYwnqZP65REXYrHOxY%2BpldrNVRYgmdInrUkInZFF5gFDoKxgHCVJTLVG0xpmOatRYdVYSsDs3w8vkCkp0pGXQE3cIuLEPM4XK5ao506no845Q3ApX8wXNEVuIA

Reply:
Yes, it is because the first case in `should_fail` raises a fast fail and shadows the following cases. Specifically, this test case should fail but won't in your solution:

```python
    def name_age(name: str, age: int) -> None:
        ...
    test_func(name_age)
```

This is caused by the current implementation of how we check "expect to fail" cases, which I plan to improve in the coming PR. And I already have some ideas for this.

Reply:
Understood. From my point of view, the semantic of `should_fail` is "it contains at least one type error", rather than "every test case should fail". Could we hold on adding this challenge until we find satisfactory test cases? We can submit the others first.

Reply:
Sometimes we have to add more than one test case to prevent users from programming based on test cases in order to take shortcuts.

Reply:
Up to you if want to continue refine the test cases or submit the others first, both work for me.

What if we use keyword-only parameters to force using Protocol?
https://docs.python.org/3.12/library/typing.html#annotating-callables:~:text=Callable%20cannot%20express%20complex%20signatures%20such%20as%20functions%20that%20take%20a%20variadic%20number%20of%20arguments%2C%20overloaded%20functions%2C%20or%20functions%20that%20have%20keyword%2Donly%20parameters.

Reply:
I decided to remove this challenge for now since I can't make a valid test case for this.